pkg/controller/direct/kms/autokeyconfig_controller.go

```diff
@@ -0,0 +1,273 @@
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type Adapter struct {
+	id        *KMSAutokeyConfigIdentity
+	gcpClient *gcp.AutokeyAdminClient
+	desired   *krm.KMSAutokeyConfig
+	actual    *kmspb.AutokeyConfig
+}
+
+var _ directbase.Adapter = &Adapter{}
+
+func (a *Adapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting KMSAutokeyConfig", "name", a.id.FullyQualifiedName())
+	req := &kmspb.GetAutokeyConfigRequest{Name: a.id.FullyQualifiedName()}
+	autokeyconfigpb, err := a.gcpClient.GetAutokeyConfig(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting KMSAutokeyConfig %q: %w", a.id.FullyQualifiedName(), err)
+	}
+
+	a.actual = autokeyconfigpb
+	return true, nil
+}
+
+func (a *Adapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	u := createOp.GetUnstructured()
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("creating AutokeyConfig", "name", a.id.FullyQualifiedName())
+	mapCtx := &direct.MapContext{}
+	desired := a.desired.DeepCopy()
+	resource := KMSAutokeyConfigSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	updated, err := a.updateAutokeyConfig(ctx, resource)
+	if err != nil {
+		return err
+	}
+	status := &krm.KMSAutokeyConfigStatus{}
+	status.ObservedState = KMSAutokeyConfigStatusObservedState_FromProto(mapCtx, updated)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = a.id.AsExternalRef()
+	return setStatus(u, status)
+}
+
+func (a *Adapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	u := updateOp.GetUnstructured()
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("updating AutokeyConfig", "name", a.id.FullyQualifiedName())
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := KMSAutokeyConfigSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	updated, err := a.updateAutokeyConfig(ctx, resource)
+	if err != nil {
+		return err
+	}
+
+	status := &krm.KMSAutokeyConfigStatus{}
+	status.ObservedState = KMSAutokeyConfigObservedState_FromProto(mapCtx, updated)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	return setStatus(u, status)
+}
+
+func (a *Adapter) updateAutokeyConfig(ctx context.Context, resource *kmspb.AutokeyConfig) (*kmspb.AutokeyConfig, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	// To populate a.actual calling a.Find()
+	isExist, err := a.Find(ctx)
+	if !isExist {
+		return nil, fmt.Errorf("updateAutokeyConfig failed as AutokeyConfig does not exist, name: %s", a.id.FullyQualifiedName())
+	}
+	if err != nil {
+		return nil, err
+	}
+	updateMask := &fieldmaskpb.FieldMask{}
+	if resource.KeyProject != "" && !reflect.DeepEqual(resource.KeyProject, a.actual.KeyProject) {
```

Comment:
you may want to compare the `.KeyProject.External` rather than the `KeyProject` itself. I think DeepEqual will always give false in that case

Reply:
I think this comment still applies 😄